### PR TITLE
ENG-447 Set weaviate index name based on copilot name

### PR DESCRIPTION
**Pull request by kristjanpeterson1**
Index name is uppercase letters from copilot name and if copilot name not set or does not contain letters then defaults to "OPENCOPILOT".
Save latest copilot configuration (copilot name, api port) to ~/.opencopilot/configuration.json
CLI uses the conf file to get weaviate index name and API port

### Task / problem
We want to be able to run multiple copilots in one machine easily. With the current solution we use a hardcoded index name and the whole schema gets deleted, which makes it impossible to run multiple copilots with the same weaviate instance.

### Potential issues
- If we fail to save copilot configuration to ~/.opencopilot/configuration.json then CLI will not work (before if you were using any other api port than 3000 for example it didnt work either btw)

### Checklist

- [X] If relevant, checked and updated the [docs repo](https://github.com/opencopilotdev/docs), [README](https://github.com/opencopilotdev/opencopilot/blob/main/README.md), and [examples](https://github.com/opencopilotdev/opencopilot/tree/main/examples)

### CODE REVIEW

**Comment by taivop** [comment]
If I start 2 copilots with the same name, they will still share the knowledge base partition?

**Comment by kristjanpeterson1** [comment]
> If I start 2 copilots with the same name, they will still share the same vector store partition?

Yes, not optimal maybe but its already a big step up from all the copilots sharing the same index in the vector store

**Comment by taivop** [comment]
Sure, makes sense.

Taivo Pungas

File: opencopilot/settings.py [hunk 1]
```diff
@@ -8,6 +12,49 @@
...
+        try:
+            return AppConf(
+                copilot_name=value["copilot_name"],
+                api_port=value["api_port"],
+            )
+        except:
+            return None
+
+    @staticmethod
+    def get() -> Optional[Any]:
```
Comment (KasparPeterson) [inline opencopilot/settings.py:40]: Should return `Optional[AppConf]` instead?

File: opencopilot/settings.py [hunk 2]
```diff
@@ -8,6 +12,49 @@
...
+    api_port: int
+
+    def to_dict(self) -> Dict:
+        return {
+            "copilot_name": self.copilot_name,
+            "api_port": self.api_port,
+        }
+
+    @staticmethod
+    def from_dict(value: Dict) -> Optional[Any]:
```
Comment (KasparPeterson) [inline opencopilot/settings.py:30]: Should return `Optional[AppConf]` instead?

File: opencopilot/settings.py [hunk 3]
```diff
@@ -8,6 +12,49 @@
...
+                    value = json.load(file)
+                return AppConf.from_dict(value)
+        except:
+            pass
+        return None
+
+    def save(self):
+        try:
+            os.makedirs(os.path.dirname(CONF_FILE_PATH), exist_ok=True)
+            with open(CONF_FILE_PATH, "w") as file:
```
Comment (KasparPeterson) [inline opencopilot/settings.py:53]: If you are already changing then maybe add encoding="utf-8" as well?

**Comment by kgrofelnik** [comment]
We should maybe consider introducing `id: UUID` parameter to OpenCopilot

**Review by taivop — APPROVED** [review]
blind approve

**Pull request merged**